src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs

```diff
@@ -1,54 +1,129 @@
 ﻿namespace NServiceBus.Persistence.CosmosDB;
 
+using System;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Extensibility;
 using Microsoft.Azure.Cosmos;
 using Newtonsoft.Json;
 using Outbox;
-using Transport;
 
-class OutboxPersister(ContainerHolderResolver containerHolderResolver, JsonSerializer serializer, int ttlInSeconds)
+class OutboxPersister(ContainerHolderResolver containerHolderResolver, JsonSerializer serializer, string partitionKey, bool readFallbackEnabled, ExtractorConfigurationHolder extractorConfigHolder, int ttlInSeconds)
     : IOutboxStorage
 {
+    readonly ExtractorConfiguration extractorConfig = extractorConfigHolder?.Configuration ?? new ExtractorConfiguration();
+
     public Task<IOutboxTransaction> BeginTransaction(ContextBag context, CancellationToken cancellationToken = default)
     {
         var cosmosOutboxTransaction = new CosmosOutboxTransaction(containerHolderResolver, context);
 
+        // Only set partition key if:
+        // 1. We have one in context AND
+        // 2. We're not going to defer to logical stage for container extraction
         if (context.TryGet(out PartitionKey partitionKey))
         {
-            cosmosOutboxTransaction.PartitionKey = partitionKey;
+            // Check if we'll defer for container extraction
+            var containerHolder = containerHolderResolver.ResolveAndSetIfAvailable(context);
+            bool willDeferForContainer = containerHolder == null && extractorConfig.HasCustomContainerMessageExtractors;
```

Comment:
Does it matter if the container is set here? ie if there is a custom extractor shouldn't we defer so that it's used?